specs/ThrottlingControlScript.md

```diff
@@ -0,0 +1,325 @@
+
+```C#
+void SetupIsolatedFramesHandler()
+{
+    // You can use the frame properties to determine whether it should be
+    // marked to be throttled separately from main frame.
+    webView.CoreWebView2.FrameCreated += (sender, args) =>
+    {
+        if (args.Frame.Name == "isolated")
+        {
+            args.Frame.ShouldUseIsolatedThrottling = true;
+        }
+    };
+
+    webView.CoreWebView2.Settings.ThrottlingIntervalPreferenceIsolated = 500;
+}
+```
+
+```cpp
+void ScenarioThrottlingControl::SetupIsolatedFramesHandler()
+{
+    auto webview4 = m_webview.try_query<ICoreWebView2_4>();
+    CHECK_FEATURE_RETURN_EMPTY(webview4);
+
+    // You can use the frame properties to determine whether it should be
+    // marked to be throttled separately from main frame.
+    CHECK_FAILURE(webview4->add_FrameCreated(
+        Callback<ICoreWebView2FrameCreatedEventHandler>(
+            [this](ICoreWebView2* sender, ICoreWebView2FrameCreatedEventArgs* args) -> HRESULT
+            {
+                wil::com_ptr<ICoreWebView2Frame> webviewFrame;
+                CHECK_FAILURE(args->get_Frame(&webviewFrame));
+
+                auto webviewFrame6 =
+                    webviewFrame.try_query<ICoreWebView2Frame6>();
+                CHECK_FEATURE_RETURN_HRESULT(webviewFrame6);
+
+                wil::unique_cotaskmem_string name;
+                CHECK_FAILURE(webviewFrame->get_Name(&name));
+                if (wcscmp(name.get(), L"isolated") == 0)
+                {
+                    CHECK_FAILURE(webviewFrame6->put_ShouldUseIsolatedThrottling(TRUE));
+                }
+
+                return S_OK;
+            })
+            .Get(),
+        &m_frameCreatedToken));
+
+    wil::com_ptr<ICoreWebView2Settings> settings;
+    m_webview->get_Settings(&settings);
+    auto settings2 = settings.try_query<ICoreWebView2Settings9>();
+    CHECK_FAILURE(settings2->put_ThrottlingIntervalPreferenceIsolated(500));
+}
+```
+
+# API Details
+```cpp
+/// A continuation of the `ICoreWebView2Settings` interface to support
+/// ThrottlingPreference.
+[uuid(00f1b5fb-91ed-4722-9404-e0f8fd1e6b0a), object, pointer_default(unique)]
+interface ICoreWebView2Settings9 : ICoreWebView2Settings8 {
+  /// The preferred wake up interval (in milliseconds) to use for throttleable
+  /// JavaScript tasks (`setTimeout` and `setInterval`), when the WebView is in
+  /// foreground state. A WebView is in foreground state when its `IsVisible`
+  /// property is `TRUE`.
+  ///
+  /// A wake up interval is the amount of time that needs to pass before the
+  /// WebView2 Runtime checks for new timer tasks to run.
+  ///
+  /// The WebView2 Runtime will try to respect the preferred interval set by the
+  /// application, but the effective value will be constrained by resource and
+  /// platform limitations. Setting a value of `0` means a preference of no
+  /// throttling to be applied. The default value is a constant determined by
```

Comment:
It looks like we're missing a way to return to the WebView2 Runtime's default value.  (In other words, there's no input for "I don't have a preference" for any of these properties?)  I might expect preference=0 means I don't have any preference, but instead it means I prefer zero throttling.

Best case, the constant determined by the running version of the WebView2 Runtime is fixed, and my app can save it before overriding if I'd care about restoring it to 'default'?

Reply:
```suggestion
  /// throttling to be applied. The default value is determined by
```

I would simplify the text to say that the default value is determined by the runtime. Don't over-promise that it's a constant. It might be determined dynamically (one value if on AC, another if on battery, and another if in energy saver mode).

It seems that the only way to restore default behavior is to save the old value and set it back, and hope that there is no dynamic throttling.

Reply:
Yes, this is a known limitation and the only solution in current design is to save the value before changes. The intention behind this was to keep the API simpler, but if this is semantically meaningful, I can think of two options:

**option 1**
  * add toggle - e.g., `CoreWebView2Settings.IsThrottlingEnabled`
    * under current pattern, it would be 4 additional properties
    * alternatively, we can bring back the "throttling category" enum and use for both interval and toggle
  * make 0 reset to default

**option 2**
  * make 0 reset to default
  * make 1-4 mean off
    * due to architectural reasons, there's always an effective minimum 4 ms throttling (even when throttling is off)

**none (current)**
  - save value before changing, then restore

... or a reversal of option 1? (method for reset and keep 0 meaning off) I think option 1 is better than its reversal.